pubsub/src/subscription.rs

```diff
@@ -340,9 +344,20 @@ impl Subscription {
     /// ```
     pub async fn subscribe(&self, opt: Option<SubscriberConfig>) -> Result<MessageStream, Status> {
         let (tx, rx) = async_channel::unbounded::<ReceivedMessage>();
-
         let cancel = CancellationToken::new();
-        Subscriber::start(cancel.clone(), self.fqsn.clone(), self.subc.clone(), tx, opt);
+
+        // spawn a separate subscriber task for each connection in the pool
+        for _ in 0..self.pool_size() {
+            let cancel = cancel.clone();
+            let fqsn = self.fqsn.clone();
+            let subc = self.subc.clone();
+            let tx = tx.clone();
+            let opt = opt.clone();
+
+            tokio::spawn(async move {
+                Subscriber::start(cancel, fqsn, subc, tx, opt);
```

Comment:
Since `tokio::spawn` is done in Subscriber::start, I think it is unnecessary `tokio::spawn` for each start, but what do you think?
https://github.com/yoshidan/google-cloud-rust/blob/7e47243374bdae3bea736431345678e31ddebc4f/pubsub/src/subscriber.rs#L118

Reply:
It does seem a bit of a waste. The reason I choose this approach was because it involved the least changes.

If the `tokio::spawn` was removed and the `for _ in 0..self.pool_size()`  was moved inside `Subscriber::start()` we would need to make changes like:
```rust
// renamed from Subscriber
struct SubscriberStream {
    pinger: TaskHandle,
    inner: TaskHandle,
}

impl SubscriberStream {
    pub async fn done(&mut self) {
        if let Some(v) = self.pinger.take() {
            let _ = v.await;
        }
        if let Some(v) = self.inner.take() {
            let _ = v.await;
        }
    }
}

struct Subscriber {
    streams: Vec<SubscriberStream>,
}

impl Subscriber {
    pub async fn done(&mut self) {
        for stream in self.stream {
            stream.done().await;
        }
    }
}
```

Reply:
What I said is that you can simply remove `tokio::spawn` and there should be no need to modify `Subscriber`.

```rust
for _ in 0..self.pool_size() {
            let cancel = cancel.clone();
            let fqsn = self.fqsn.clone();
            let subc = self.subc.clone();
            let tx = tx.clone();
            let opt = opt.clone();
            Subscriber::start(cancel, fqsn, subc, tx, opt);
} 
```

I've tested publishing 100 messages it with both codes and there is some bias at first, but both codes seem to process messages equally well.


1. No `tokio::spawn`
```rust
for i in 0..self.pool_size() {
            let cancel = cancel.clone();
            let fqsn = self.fqsn.clone();
            let subc = self.subc.clone();
            let tx = tx.clone();
            let opt = opt.clone();
            Subscriber::start(i, cancel, fqsn, subc, tx, opt);
} 
```

(The subscriber's number is logged for confirmation.)
```
2023-05-26T02:16:22.890407Z TRACE google_cloud_pubsub::subscriber: start subscriber: 3 : projects/atl-dev1/subscriptions/test-subscription
2023-05-26T02:16:22.890407Z TRACE google_cloud_pubsub::subscriber: start subscriber: 1 : projects/atl-dev1/subscriptions/test-subscription
2023-05-26T02:16:22.890421Z TRACE google_cloud_pubsub::subscriber: start subscriber: 2 : projects/atl-dev1/subscriptions/test-subscription
2023-05-26T02:16:22.890461Z TRACE google_cloud_pubsub::subscriber: start subscriber: 0 : projects/atl-dev1/subscriptions/test-subscription
2023-05-26T02:16:24.682886Z TRACE google_cloud_pubsub::subscriber: start streaming: 2 projects/atl-dev1/subscriptions/test-subscription
2023-05-26T02:16:24.682984Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811049176330543
2023-05-26T02:16:26.981557Z TRACE google_cloud_pubsub::subscriber: start streaming: 3 projects/atl-dev1/subscriptions/test-subscription
2023-05-26T02:16:27.001041Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811062911367417
2023-05-26T02:16:27.001139Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811049943351492
2023-05-26T02:16:27.001154Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811074177414518
2023-05-26T02:16:27.001166Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811008299510143
2023-05-26T02:16:27.001177Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811061308750791
2023-05-26T02:16:27.002276Z TRACE google_cloud_pubsub::subscriber: start streaming: 1 projects/atl-dev1/subscriptions/test-subscription
2023-05-26T02:16:27.002385Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7976602536536280
2023-05-26T02:16:27.002416Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811049895573999
2023-05-26T02:16:27.004085Z TRACE google_cloud_pubsub::subscriber: start streaming: 0 projects/atl-dev1/subscriptions/test-subscription
2023-05-26T02:16:27.004158Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811034153626604
2023-05-26T02:16:27.011330Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811034412795185
2023-05-26T02:16:27.071869Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811033595502706
2023-05-26T02:16:27.129503Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811049646568925
2023-05-26T02:16:27.198143Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811061683310103
2023-05-26T02:16:27.257516Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811061417855447
2023-05-26T02:16:27.319647Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811062576345895
2023-05-26T02:16:27.384351Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811062548087245
2023-05-26T02:16:27.453939Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811020384032701
2023-05-26T02:16:27.509214Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811062218794438
2023-05-26T02:16:27.581094Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811033836296469
2023-05-26T02:16:27.640298Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811048474654984
2023-05-26T02:16:27.704050Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811034036645321
2023-05-26T02:16:27.770540Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811049659085109
2023-05-26T02:16:27.833190Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811033667147565
2023-05-26T02:16:27.896075Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811034167199966
2023-05-26T02:16:27.957587Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811063067481527
2023-05-26T02:16:27.974030Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811033595498887
2023-05-26T02:16:28.034289Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811033939649183
2023-05-26T02:16:28.092824Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811020864045870
2023-05-26T02:16:28.150218Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811021486214810
2023-05-26T02:16:28.206214Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811074005991368
2023-05-26T02:16:28.224302Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811008299501540
2023-05-26T02:16:28.285733Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811050085360449
2023-05-26T02:16:28.305765Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811061855469603
2023-05-26T02:16:28.359302Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811020944074814
2023-05-26T02:16:28.427072Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811061138680184
2023-05-26T02:16:28.485613Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7810995255566892
2023-05-26T02:16:28.536234Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811009434914112
2023-05-26T02:16:28.599480Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811074187831803
2023-05-26T02:16:28.662646Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811034075986929
2023-05-26T02:16:28.679934Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811061683305903
2023-05-26T02:16:28.695856Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811061417851565
2023-05-26T02:16:28.714301Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811049895555736
2023-05-26T02:16:28.854502Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811049853085789
2023-05-26T02:16:28.969927Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811033860167109
2023-05-26T02:16:29.038371Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811035009164842
2023-05-26T02:16:29.166630Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811009187811750
2023-05-26T02:16:29.239156Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811021755242343
2023-05-26T02:16:29.258124Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811073953853127
2023-05-26T02:16:29.386949Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811049177090050
2023-05-26T02:16:29.456316Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811034061446581
2023-05-26T02:16:29.480165Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811033211363112
2023-05-26T02:16:29.537400Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811033463650565
2023-05-26T02:16:29.550650Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811049895554056
2023-05-26T02:16:29.641257Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7810983174141423
2023-05-26T02:16:29.703013Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811074867287150
2023-05-26T02:16:29.764470Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811048189062510
2023-05-26T02:16:29.834648Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811074315643737
2023-05-26T02:16:29.892757Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811034052191032
2023-05-26T02:16:29.942864Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811061874378943
2023-05-26T02:16:29.997816Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811048235986667
2023-05-26T02:16:30.055610Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811034170016691
2023-05-26T02:16:30.073366Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811034636445045
2023-05-26T02:16:30.135177Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811074355231690
2023-05-26T02:16:30.196217Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811034404406047
2023-05-26T02:16:30.223082Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811033874310856
2023-05-26T02:16:30.239427Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811020864040890
2023-05-26T02:16:30.313672Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811062645138620
2023-05-26T02:16:30.326832Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811075374443227
2023-05-26T02:16:30.393298Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811009192449738
2023-05-26T02:16:30.448524Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811035199466699
2023-05-26T02:16:30.470351Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811074837733574
2023-05-26T02:16:30.485818Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811062576539298
2023-05-26T02:16:30.549501Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811020464575996
2023-05-26T02:16:30.615312Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811075058339908
2023-05-26T02:16:30.628100Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811034534315041
2023-05-26T02:16:30.708446Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811034209322778
2023-05-26T02:16:30.720109Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7976275231428909
2023-05-26T02:16:30.736945Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811074100208646
2023-05-26T02:16:30.804978Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811061314089148
2023-05-26T02:16:30.819212Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811061314089134
2023-05-26T02:16:30.883630Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811034268490191
2023-05-26T02:16:30.899772Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811021486209006
2023-05-26T02:16:30.921694Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811073953849190
2023-05-26T02:16:30.979212Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811020819994946
2023-05-26T02:16:31.001717Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811034660300460
2023-05-26T02:16:31.061605Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811049973794169
2023-05-26T02:16:31.076771Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811074315640482
2023-05-26T02:16:31.141488Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811033709703203
2023-05-26T02:16:31.210831Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811062981521475
2023-05-26T02:16:31.228121Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811049657719091
2023-05-26T02:16:31.248452Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811061811814809
2023-05-26T02:16:31.268058Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811061138674805
2023-05-26T02:16:31.328928Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811034560991512
2023-05-26T02:16:31.389425Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811050278229090
2023-05-26T02:16:31.406176Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7976602536516108
2023-05-26T02:16:31.461007Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811074705691462
2023-05-26T02:16:31.479099Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811074100207092
2023-05-26T02:16:31.532838Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811049888909011
2023-05-26T02:16:31.548211Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811061308732954
2023-05-26T02:16:31.568271Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7976320947817868
```

2. with `tokio::spawn`
```rust
for i in 0..self.pool_size() {
            let cancel = cancel.clone();
            let fqsn = self.fqsn.clone();
            let subc = self.subc.clone();
            let tx = tx.clone();
            let opt = opt.clone();
            tokio::spawn(async move {
              Subscriber::start(i, cancel, fqsn, subc, tx, opt);
           });
} 
```
```
2023-05-26T02:17:17.759768Z TRACE google_cloud_pubsub::subscriber: start subscriber: 3 : projects/atl-dev1/subscriptions/test-subscription
2023-05-26T02:17:17.759959Z TRACE google_cloud_pubsub::subscriber: start subscriber: 2 : projects/atl-dev1/subscriptions/test-subscription
2023-05-26T02:17:17.759758Z TRACE google_cloud_pubsub::subscriber: start subscriber: 1 : projects/atl-dev1/subscriptions/test-subscription
2023-05-26T02:17:17.760751Z TRACE google_cloud_pubsub::subscriber: start subscriber: 0 : projects/atl-dev1/subscriptions/test-subscription
2023-05-26T02:17:19.002851Z TRACE google_cloud_pubsub::subscriber: start streaming: 2 projects/atl-dev1/subscriptions/test-subscription
2023-05-26T02:17:19.002962Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811062870112562
2023-05-26T02:17:19.002993Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811034561067940
2023-05-26T02:17:19.003006Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811061874494368
2023-05-26T02:17:19.004201Z TRACE google_cloud_pubsub::subscriber: start streaming: 0 projects/atl-dev1/subscriptions/test-subscription
2023-05-26T02:17:19.004265Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811034061697774
2023-05-26T02:17:19.004924Z TRACE google_cloud_pubsub::subscriber: start streaming: 3 projects/atl-dev1/subscriptions/test-subscription
2023-05-26T02:17:19.005063Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811048391014876
2023-05-26T02:17:19.005080Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811062218880461
2023-05-26T02:17:19.005093Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811062578397842
2023-05-26T02:17:19.005105Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811074188054520
2023-05-26T02:17:19.005118Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811074705784119
2023-05-26T02:17:19.005248Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811035009242022
2023-05-26T02:17:19.005897Z TRACE google_cloud_pubsub::subscriber: start streaming: 1 projects/atl-dev1/subscriptions/test-subscription
2023-05-26T02:17:19.005939Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811100137296751
2023-05-26T02:17:19.006248Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811033640834880
2023-05-26T02:17:19.006281Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811074837808060
2023-05-26T02:17:19.065195Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811049895796585
2023-05-26T02:17:19.122887Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811062981558672
2023-05-26T02:17:19.187435Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811074036876883
2023-05-26T02:17:19.248192Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811062744968175
2023-05-26T02:17:19.315031Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811009255182472
2023-05-26T02:17:19.380930Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7810983174276404
2023-05-26T02:17:19.437113Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811061423959100
2023-05-26T02:17:19.522275Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811074867351274
2023-05-26T02:17:19.600131Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811075374639258
2023-05-26T02:17:19.659112Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811033596831849
2023-05-26T02:17:19.687873Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811034061696491
2023-05-26T02:17:19.697806Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811020464630010
2023-05-26T02:17:19.756855Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811020944339269
2023-05-26T02:17:19.824446Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811034875578517
2023-05-26T02:17:19.848923Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811062218877915
2023-05-26T02:17:19.857949Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811074036874548
2023-05-26T02:17:19.916724Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811033836304742
2023-05-26T02:17:19.980712Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811062382033244
2023-05-26T02:17:20.062032Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811061175645836
2023-05-26T02:17:20.121788Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811022362375415
2023-05-26T02:17:20.185793Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811034209410530
2023-05-26T02:17:20.245494Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811074006080980
2023-05-26T02:17:20.264678Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811049895792856
2023-05-26T02:17:20.277780Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811061424157817
2023-05-26T02:17:20.296727Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811074177459044
2023-05-26T02:17:20.318792Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811061314176351
2023-05-26T02:17:20.391366Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811062716188372
2023-05-26T02:17:20.446535Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811049902094924
2023-05-26T02:17:20.464286Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811074188048051
2023-05-26T02:17:20.480241Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811033667221101
2023-05-26T02:17:20.547943Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811061138761147
2023-05-26T02:17:20.598693Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811033781636977
2023-05-26T02:17:20.614986Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811049176597876
2023-05-26T02:17:20.632834Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811033927266775
2023-05-26T02:17:20.658274Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811033927266745
2023-05-26T02:17:20.728382Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811049853115265
2023-05-26T02:17:20.791233Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811020828680684
2023-05-26T02:17:20.855992Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811033709806943
2023-05-26T02:17:20.886732Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811034052235316
2023-05-26T02:17:20.899005Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811061175644373
2023-05-26T02:17:20.961883Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811073953908876
2023-05-26T02:17:20.983321Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811049465069640
2023-05-26T02:17:21.018215Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811050286178533
2023-05-26T02:17:21.076970Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811034134940469
2023-05-26T02:17:21.096497Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811034052234829
2023-05-26T02:17:21.173234Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811048315134567
2023-05-26T02:17:21.200406Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811061308802385
2023-05-26T02:17:21.213258Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811061314174953
2023-05-26T02:17:21.281917Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811033874411489
2023-05-26T02:17:21.348084Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811082591757710
2023-05-26T02:17:21.409436Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811021536004598
2023-05-26T02:17:21.428602Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811034875576087
2023-05-26T02:17:21.444730Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811050085421925
2023-05-26T02:17:21.459305Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811035199707468
2023-05-26T02:17:21.477825Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811020464626575
2023-05-26T02:17:21.493479Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811100137284526
2023-05-26T02:17:21.509702Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811074003851114
2023-05-26T02:17:21.531512Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811033738427166
2023-05-26T02:17:21.599073Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811049657783127
2023-05-26T02:17:21.665927Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811062138421838
2023-05-26T02:17:21.682207Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811074867345980
2023-05-26T02:17:21.695847Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811021536004048
2023-05-26T02:17:21.763054Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7976602536792177
2023-05-26T02:17:21.776858Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7976602536792156
2023-05-26T02:17:21.844083Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7810995477445953
2023-05-26T02:17:21.906259Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811075028817630
2023-05-26T02:17:21.985824Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811034506632281
2023-05-26T02:17:22.040508Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811062857962712
2023-05-26T02:17:22.098957Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7976456152288099
2023-05-26T02:17:22.126054Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811074036868469
2023-05-26T02:17:22.147549Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811062122160337
2023-05-26T02:17:22.206921Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811033860218037
2023-05-26T02:17:22.224826Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811061874489152
2023-05-26T02:17:22.284039Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811034441426694
2023-05-26T02:17:22.306279Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811033738426434
2023-05-26T02:17:22.359436Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811061855542494
2023-05-26T02:17:22.430693Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811049646664734
2023-05-26T02:17:22.447054Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7976602536790328
2023-05-26T02:17:22.462639Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811050085419374
2023-05-26T02:17:22.531858Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811020384278184
2023-05-26T02:17:22.589535Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811061990301102
2023-05-26T02:17:22.607236Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811062138419075
2023-05-26T02:17:22.628526Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811062981548997
2023-05-26T02:17:22.645668Z DEBUG google_cloud_pubsub::subscriber: message received: 0 : msg_id=7811009392318269
2023-05-26T02:17:22.705325Z DEBUG google_cloud_pubsub::subscriber: message received: 3 : msg_id=7811074315891195
2023-05-26T02:17:22.721390Z DEBUG google_cloud_pubsub::subscriber: message received: 1 : msg_id=7811034660358236
2023-05-26T02:17:22.750580Z DEBUG google_cloud_pubsub::subscriber: message received: 2 : msg_id=7811063069954442
```

Reply:
Ahh apologies, I understand now. I will make the change you suggested.

> For example, currently subscriber A will use connection 0, 1, 2, 3, 0, ... and subscriber B will use connection 1, 2, 3, 0, 1, .. for requests. I wonder if it might be better for subscriber A to use just connection 0 and subscriber B to use just connection 1, etc (assuming there is the same number of connections and subscriber clients.

In this comment of mine, I was trying to talk about how to solve this bias (of different subscribers trying to use the same connection), so it's interesting to see that the set of subscribers sort themselves out after some time. I assume this is due to multiple subscribers trying to use a single single connection will block and only allow sequential progress. 

Example (t=time, sub=subscriber, conn=connection):
```
t=0) sub(0)=conn(0) | sub(1)=conn(0) | sub(2)=conn(0) // conn(0) has requests from sub(0), sub(1), sub(2) queued
t=1) sub(0)=conn(1) | sub(1)=conn(0) | sub(2)=conn(0) // conn(0) has requests from sub(1), sub(2) queued
t=2) sub(0)=conn(2) | sub(1)=conn(1) | sub(2)=conn(0) // conn(0) has requests from sub(2) queued
t=3) sub(0)=conn(0) | sub(1)=conn(2) | sub(2)=conn(1)
t=4) sub(0)=conn(1) | sub(1)=conn(0) | sub(2)=conn(2)
```

Hard to show but I see this effect slightly when running a benchmark locally and in some Grafana metrics.

Reply:
> In this comment of mine, I was trying to talk about how to solve this bias (of different subscribers trying to use the same connection)

Assigning a separate gRPC connection to each Subscriber will speed up the process and almost eliminate bias, so I think the method you mentioned in your PR is fine.
However, for subscriptions with message ordering enabled, bias may occur depending on the type of `ordering_key`. In the extreme case, for example, if `ordering_key="A"` for all messages, only one of the four Subscriber will be used.
Therefore, it may be better to allow only the number of Subscribers to be activated as specified in Config. (The maximum number is set as pool_size)

```rust
// same as ReceiveConfig
#[derive(Debug, Clone)]
pub struct SubscribeConfig {
    pub worker_count: usize,
    pub subscriber_config: SubscriberConfig,
}

impl Default for SubscribeConfig {
    fn default() -> Self {
        Self {
            worker_count: 1,
            subscriber_config: SubscriberConfig::default(),
        }
    }
}

pub async fn subscribe(&self, opt: Option<SubscribeConfig>) -> Result<MessageStream, Status>  {
   ...
   // default SubscribeConfig.subscriber_count = 1
   let subscribers = min(opt.unwrap_or_default.subscriber_count,  self.pool_size)
   for _ in 0..subscribers {
        Subscriber::start(...);
   }
}
```

>  I assume this is due to multiple subscribers trying to use a single single connection will block and only allow sequential progress.

The initialization bias in the [log described earlier](https://github.com/yoshidan/google-cloud-rust/pull/153#discussion_r1206184669) occurs even if the Subscriber uses a separate gRPC connection for each.
It was due to the fact that I was receiving messages from the Subscriber who started first, before all Subscriber started steaming-pull. If I wait for all Subscriber to start steaming-pull and then publish, I can confirm that the bias is eliminated as follows.
```
2023-05-26T09:05:47.238832Z TRACE google_cloud_pubsub::subscriber: start subscriber: 1 : projects/atl-dev1/subscriptions/test-subscription
2023-05-26T09:05:47.239002Z TRACE google_cloud_pubsub::subscriber: start subscriber: 2 : projects/atl-dev1/subscriptions/test-subscription
2023-05-26T09:05:47.238835Z TRACE google_cloud_pubsub::subscriber: start subscriber: 3 : projects/atl-dev1/subscriptions/test-subscription
2023-05-26T09:05:47.239252Z TRACE google_cloud_pubsub::subscriber: start subscriber: 0 : projects/atl-dev1/subscriptions/test-subscription
2023-05-26T09:05:52.988226Z TRACE google_cloud_pubsub::subscriber: start streaming: 2 projects/atl-dev1/subscriptions/test-subscription
2023-05-26T09:05:52.988494Z DEBUG google_cloud_pubsub::subscriber: message received: 2 msg_id=7814142368062474 data=0
2023-05-26T09:05:53.061733Z TRACE google_cloud_pubsub::subscriber: start streaming: 1 projects/atl-dev1/subscriptions/test-subscription
2023-05-26T09:05:53.061840Z DEBUG google_cloud_pubsub::subscriber: message received: 1 msg_id=7814092259447058 data=1
2023-05-26T09:05:53.132791Z TRACE google_cloud_pubsub::subscriber: start streaming: 3 projects/atl-dev1/subscriptions/test-subscription
2023-05-26T09:05:53.132904Z DEBUG google_cloud_pubsub::subscriber: message received: 3 msg_id=7814141650177840 data=2
2023-05-26T09:05:53.212038Z TRACE google_cloud_pubsub::subscriber: start streaming: 0 projects/atl-dev1/subscriptions/test-subscription
2023-05-26T09:05:53.212132Z DEBUG google_cloud_pubsub::subscriber: message received: 0 msg_id=7814092499430882 data=3
2023-05-26T09:05:53.282071Z DEBUG google_cloud_pubsub::subscriber: message received: 2 msg_id=7814200872266291 data=4
2023-05-26T09:05:53.352793Z DEBUG google_cloud_pubsub::subscriber: message received: 1 msg_id=7814092884362154 data=5
2023-05-26T09:05:53.423214Z DEBUG google_cloud_pubsub::subscriber: message received: 3 msg_id=7814140617835431 data=6
2023-05-26T09:05:53.498226Z DEBUG google_cloud_pubsub::subscriber: message received: 0 msg_id=7814141028290707 data=7
2023-05-26T09:05:53.521640Z DEBUG google_cloud_pubsub::subscriber: message received: 2 msg_id=7814092884361555 data=8
2023-05-26T09:05:53.588685Z DEBUG google_cloud_pubsub::subscriber: message received: 1 msg_id=7814141265395991 data=9
```

If a large number of messages have already accumulated in subscriptions, it would be difficult to eliminate the bias during initialization.